src/keys/keypair.rs

```diff
@@ -0,0 +1,241 @@
+        }
+    }
+
+    /// Signs with the underlying keypair.
+    pub fn sign(&self, data: &[u8]) -> Signature {
+        match self {
+            Self::Ed25519(keypair) => Signature::Ed25519(keypair.sign(&data)),
+            Self::Bls(keypair) => Signature::Bls(keypair.secret.sign(data)),
+            Self::BlsShare(keypair) => {
+                let index = keypair.index;
+                let share = keypair.secret.sign(data);
+                Signature::BlsShare(SignatureShare { index, share })
+            }
+        }
+    }
+}
+
+impl From<threshold_crypto::SecretKey> for Keypair {
+    fn from(sk: threshold_crypto::SecretKey) -> Self {
+        let public = sk.public_key();
+        let keypair = BlsKeypair {
+            secret: SerdeSecret(sk),
+            public,
+        };
+        Self::Bls(keypair)
+    }
+}
+
+impl From<SerdeSecret<threshold_crypto::SecretKey>> for Keypair {
+    fn from(sk: SerdeSecret<threshold_crypto::SecretKey>) -> Self {
+        let public = sk.public_key();
+        let keypair = BlsKeypair { secret: sk, public };
+        Self::Bls(keypair)
+    }
+}
+
+impl From<ed25519_dalek::SecretKey> for Keypair {
+    fn from(secret: ed25519_dalek::SecretKey) -> Self {
+        let public: ed25519_dalek::PublicKey = (&secret).into();
+
+        let keypair = ed25519_dalek::Keypair { public, secret };
+
+        Self::Ed25519(keypair)
+    }
+}
+
+/// BLS keypair.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct BlsKeypair {
+    /// Secret key.
+    pub secret: SerdeSecret<threshold_crypto::SecretKey>,
+    /// Public key.
+    pub public: threshold_crypto::PublicKey,
+}
+
+/// BLS keypair share.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct BlsKeypairShare {
+    /// Share index.
+    pub index: usize,
+    /// Secret key share.
+    pub secret: SerdeSecret<threshold_crypto::SecretKeyShare>,
+    /// Public key share.
+    pub public: threshold_crypto::PublicKeyShare,
+    /// Public key set. Necessary for producing proofs.
+    pub public_key_set: threshold_crypto::PublicKeySet,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::utils;
+    use bincode::deserialize as deserialise;
+    use threshold_crypto::{self};
+
+    fn gen_keypairs() -> Vec<Keypair> {
+        let mut rng = rand::thread_rng();
+        let bls_secret_key = threshold_crypto::SecretKeySet::random(1, &mut rng);
+        vec![
+            Keypair::new_ed25519(&mut rng),
+            Keypair::new_bls(&mut rng),
+            Keypair::new_bls_share(
+                0,
+                bls_secret_key.secret_key_share(0),
+                bls_secret_key.public_keys(),
+            ),
+        ]
+    }
+
+    fn gen_keys() -> Vec<PublicKey> {
+        gen_keypairs().iter().map(PublicKey::from).collect()
+    }
+
+    // Test serialising and deserialising key pairs.
+    #[test]
+    fn serialisation_key_pair() {
```

Comment:
if we make this a `Result<>` we can use a `?` and not unwrap (we need to kill that crate anyway. If we need unwrap we can use the std `.unwrap("comment")`